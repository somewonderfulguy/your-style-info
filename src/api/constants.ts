export const pagePathPair = new Map([
  ['/en', '/json/home-en.json'],
  ['/ru', '/json/home-ru.json'],
  ['/en/guides-topics', '/json/guides-en.json'],
  ['/ru/guides-topics', '/json/guides-ru.json'],
  ['/en/guides-topics/colors-how-to', '/json/colors-how-to-en.json'],
  ['/ru/guides-topics/colors-how-to', '/json/colors-how-to-ru.json'],
  ['/en/guides-topics/how-to-travel', '/json/how-to-travel-en.json'],
  ['/ru/guides-topics/how-to-travel', '/json/how-to-travel-ru.json'],
  ['/en/guides-topics/black-white-ties', '/json/black-white-ties-en.json'],
  ['/ru/guides-topics/black-white-ties', '/json/black-white-ties-ru.json'],
  ['/en/clothes', '/json/clothes-en.json'],
  ['/ru/clothes', '/json/clothes-ru.json'],
  ['/en/clothes/suits-jackets-vests', '/json/suits-jackets-vests-en.json'],
  ['/ru/clothes/suits-jackets-vests', '/json/suits-jackets-vests-ru.json'],
  ['/en/clothes/shirts', '/json/shirts-en.json'],
  ['/ru/clothes/shirts', '/json/shirts-ru.json'],
  ['/en/clothes/knitwear', '/json/knitwear-en.json'],
  ['/ru/clothes/knitwear', '/json/knitwear-ru.json'],
  ['/en/clothes/shoes', '/json/shoes-en.json'],
  ['/ru/clothes/shoes', '/json/shoes-ru.json'],
  ['/en/clothes/outerwear', '/json/outerwear-en.json'],
  ['/ru/clothes/outerwear', '/json/outerwear-ru.json'],
  ['/en/clothes/accessories', '/json/accessories-en.json'],
  ['/ru/clothes/accessories', '/json/accessories-ru.json'],
  ['/en/clothes/materials-and-fabrics', '/json/materials-and-fabrics-en.json'],
  ['/ru/clothes/materials-and-fabrics', '/json/materials-and-fabrics-ru.json'],
  ['/en/clothes/patterns', '/json/patterns-en.json'],
  ['/ru/clothes/patterns', '/json/patterns-ru.json'],
  ['/en/clothes/maintenance', '/json/maintenance-en.json'],
  ['/ru/clothes/maintenance', '/json/maintenance-ru.json'],
  ['/en/grooming', '/json/grooming-en.json'],
  ['/ru/grooming', '/json/grooming-ru.json'],
  ['/en/grooming/face-skin-care', '/json/face-skin-care-en.json'],
  ['/ru/grooming/face-skin-care', '/json/face-skin-care-ru.json'],
  ['/en/grooming/hands-care', '/json/hands-care-en.json'],
  ['/ru/grooming/hands-care', '/json/hands-care-ru.json'],
  ['/en/brands-stores', '/json/brands-stores-en.json'],
  ['/ru/brands-stores', '/json/brands-stores-ru.json'],
  ['/en/glossary', '/json/glossary-en.json'],
  ['/ru/glossary', '/json/glossary-ru.json']
])
