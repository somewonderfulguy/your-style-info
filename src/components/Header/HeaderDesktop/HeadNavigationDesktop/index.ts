export * from './HeadNavigationDesktop'
export { default } from './HeadNavigationDesktop'
