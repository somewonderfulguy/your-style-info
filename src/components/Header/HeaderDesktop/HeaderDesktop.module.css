.aboveNav {
  display: flex;
  justify-content: space-between;
}

.hgroup {
  max-width: 1400px;
  padding-left: 100px;
  margin: 40px 0;
}

.title {
  font-family: 'Inconsolata', monospace;
  text-transform: uppercase;
  margin: 0;
}

.subtitle {
  font-family: arial;
  font-size: 15px;
  color: darkgray;
  margin: 0;
}

.sideControlsContainer {
  margin: 18px 30px 0 0;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
}

.sideControlsSocialMedia {
  margin-top: 10px;
}

.sideControlsDarkTheme {
  margin-top: 11px;
}

.navContainer {
  position: absolute;
  left: 0;
  right: 0;
  background-color: rgba(var(--primaryBackground), .95);
  transition-duration: .2s;
}

.fixedNavContainer {
  composes: navContainer;
  position: fixed;
  top: -1px;
}