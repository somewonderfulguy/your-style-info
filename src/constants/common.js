export const SCROLL_TOP_DURATION = 250

export const STATUS = {
  idle: 'idle',
  pending: 'pending',
  loaded: 'loaded',
  error: 'error'
}