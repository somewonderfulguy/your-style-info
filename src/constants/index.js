export * from './common'
export * from './componentTypes'
export * from './languages'
export * from './primeRoutes'