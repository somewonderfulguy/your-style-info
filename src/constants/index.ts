export * from './common'
export * from './languages'
export * from './rootMenuThumbs'
export * from './primeRoutes'