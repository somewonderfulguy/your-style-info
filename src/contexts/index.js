export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'