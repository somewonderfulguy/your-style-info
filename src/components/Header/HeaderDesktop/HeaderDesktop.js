import React, {useEffect, useRef, useState} from 'react'
import {animated, useSpring, useTransition} from 'react-spring'

import {useAnimatedAppearing} from './hooks'
import {useStickyNavBar} from '../hooks'
import HeadNavigationDesktop from './HeadNavigationDesktop'
import DarkThemeSwitcher from 'components/DarkThemeSwitcher'
import LangSelector from 'components/LangSelector'
import SocialMediaIcons from 'components/SocialMediaIcons'
import {useLocalisation} from 'contexts'
import styles from './HeaderDesktop.module.css'

const HeaderDesktop = () => {
  const {translations} = useLocalisation()
  const {appearingSprings: [firstAppearing, secondAppearing, thirdAppearing, forthAppearing]} = useAnimatedAppearing()

  const navBarDOM = useRef(null)
  const headerDOM = useRef(null)

  const navbarHeight = navBarDOM.current && (navBarDOM.current.offsetHeight || 0)
  const [navBarTopLine, setNavBarTopLine] = useState(0)

  useEffect(() => {
    setNavBarTopLine(headerDOM.current && headerDOM.current.offsetHeight)
  }, [setNavBarTopLine])

  // show on scroll-up logic
  const [isRootMenuOpen, setRootMenuOpen] = useState(false)
  const [persistRootMenu, setPersistRootMenu] = useState(false)
  const navBarState = useStickyNavBar(navBarTopLine + (persistRootMenu ? 0 : navbarHeight), navBarTopLine)
  const {isFixed, isShown} = navBarState
  useEffect(() => void setRootMenuOpen(false), [navBarState])

  // animate menu on scroll up
  const {top: menuTop} = useSpring({
    config: {duration: 200},
    top: isShown || isRootMenuOpen || persistRootMenu ? -1 : navbarHeight * -1
  })

  const subtitleSelector = 'subtitle'
  const transitions = useTransition(translations && translations.subtitle, null, {
    from: {opacity: 0},
    enter: {opacity: 1},
    leave: {opacity: 0}
  })

  return (
    <header
      className={styles.header}
      style={{paddingBottom: navbarHeight}}
      ref={headerDOM}
    >
      <div className={styles.aboveNav}>
        <div className={styles.hgroup}>
          <animated.h1 className={styles.title} style={firstAppearing}>
            Your Style
          </animated.h1>
          <animated.p className={styles.subtitleContainer} style={secondAppearing}>
            {transitions.map(({item, key, props}) => (
              item !== subtitleSelector && (
                <animated.span key={key} style={props} className={styles.subtitle}>
                  {item}
                </animated.span>
              )
            ))}
          </animated.p>
        </div>
        <div className={styles.sideControlsContainer}>
          <animated.div style={{
            opacity: secondAppearing.opacity,
            transform: secondAppearing.reverseTransform,
            zIndex: 1
          }}>
            <LangSelector />
          </animated.div>
          <animated.div className={styles.sideControlsSocialMedia} style={{
            opacity: thirdAppearing.opacity,
            transform: thirdAppearing.reverseTransform
          }}>
            <SocialMediaIcons small color="lightgray" />
          </animated.div>
          <animated.div className={styles.sideControlsDarkTheme} style={{
            opacity: forthAppearing.opacity,
            transform: forthAppearing.reverseTransform
          }}>
            <DarkThemeSwitcher />
          </animated.div>
        </div>
      </div>

      <animated.nav
        ref={navBarDOM}
        style={{
          opacity: forthAppearing.opacity,
          top: isFixed ? menuTop : 'initial'
        }}
        className={isFixed ? styles.fixedNavContainer : styles.navContainer}
        onMouseEnter={() => setPersistRootMenu(true)}
        onMouseLeave={() => setPersistRootMenu(false)}
      >
        <HeadNavigationDesktop setRootMenuOpen={setRootMenuOpen} setPersistRootMenu={setPersistRootMenu} />
      </animated.nav>
    </header>
  )
}

export default HeaderDesktop