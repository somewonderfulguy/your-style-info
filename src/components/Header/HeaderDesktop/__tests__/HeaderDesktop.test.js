import React from 'react'
import {MemoryRouter} from 'react-router-dom'
import {fireEvent, render} from '@testing-library/react'
import {renderHook} from '@testing-library/react-hooks'
import user from '@testing-library/user-event'

import HeaderDesktop from '..'
import {ThemeProvider, useTheme} from 'contexts'
import { PRIME_ROUTES } from 'constants/index'

// TODO remove all waits once react-spring 9.0.0 released

const setup = () => render(<HeaderDesktop />, {
  wrapper: props => (
    <MemoryRouter>
      <ThemeProvider {...props} />
    </MemoryRouter>
  )
})

test('renders headers and social media icons', () => {
  const {getByText, getByTitle, container} = setup()
  const header = getByText(/your style/i)
  const subtitle = container.querySelector('p')
  const socialMedia = [/instagram/i, /facebook/i, /twitter/i, /vkontakte/i, /youtube/i]

  // TODO fix 18next
  expect(header).toBeInTheDocument()
  expect(subtitle).toMatchSnapshot()
  socialMedia.forEach(icon => expect(getByTitle(icon)).toBeInTheDocument())
})

test('menu (navigation) works as expected', () => {
  const {getByText} = setup()
  // getByRole, getAllByRole,
  // const navigation = getByRole('navigation')
  // const rootMenu = getAllByRole('list')[0]
  // const getSubMenu = () => getAllByRole('list')[1]

  // root menu items check
  const expectedRootMenuItems = Object.values(PRIME_ROUTES).map(o => o.name)
  const expectedRootDisabledItems = Object.values(PRIME_ROUTES).filter(o => o.inactive).map(o => o.name)

  expectedRootMenuItems.forEach(item => expect(getByText(item)).toBeInTheDocument())
  expectedRootDisabledItems.forEach(disabledItem => (
    expect(getByText(disabledItem)).toHaveAttribute('aria-disabled', 'true')
  ))

  // mouse enter
  // mouse out
  // mouse out but enter submenu

  // loop all sub items
  // in each submenu check - 1) list items, 2) disabled items, 3) img

  expect(null).toBeNull()
})

test('theme switching works as expected', () => {
  const {getByRole} = setup()
  const getThemeHook = () => renderHook(() => useTheme(), {
    wrapper: ThemeProvider
  })
  const themeSwitcher = getByRole('checkbox')

  expect(getThemeHook().result.current.isDarkTheme).toBeFalsy()

  user.click(themeSwitcher)
  expect(getThemeHook().result.current.isDarkTheme).toBeTruthy()

  user.click(themeSwitcher)
  expect(getThemeHook().result.current.isDarkTheme).toBeFalsy()
})

test('language switcher works as expected', () => {
  const { getByLabelText, getByRole, queryByRole } = setup()
  const langSelectorBtn = getByLabelText(/switch language/i)

  expect(queryByRole('menu')).toBeNull()

  // open
  user.click(langSelectorBtn)
  expect(getByRole('menu')).toBeInTheDocument()

  // close
  user.click(langSelectorBtn)
  expect(queryByRole('menu')).toBeNull()

  // open
  user.click(langSelectorBtn)
  expect(getByRole('menu')).toBeInTheDocument()

  // TODO check language switching

  // close by clicking outside
  fireEvent.click(document)
  expect(queryByRole('menu')).toBeNull()
})