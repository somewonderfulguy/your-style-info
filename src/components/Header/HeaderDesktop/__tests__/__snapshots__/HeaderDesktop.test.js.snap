// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders headers and social media icons 1`] = `
<p
  class="subtitle"
  style="opacity: 0; transform: translate3d(-20px, 0, 0);"
>
  subtitle
</p>
`;
